addons/mail/controllers/google_translate.py

```diff
@@ -0,0 +1,49 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request, route, Controller
+
+import babel
+import requests
+
+
+class GoogleTranslateController(Controller):
+    @route("/mail/message/translate", type="json", auth="user")
+    def translate(self, message_id):
+        message = request.env["mail.message"].browse(message_id)
+        message.ensure_one()
+        message._validate_access_for_current_persona("read")
+        MessageTranslation = request.env["mail.message.translation"].sudo()
+        user_lang = request.env.user.lang.split("_")[0]
+        translation = MessageTranslation.search([("message_id", "=", message_id), ("target_lang", "=", user_lang)])
+        if not translation:
+            try:
+                body = str(message.body)
+                code = self._detect_source(body)
```

Comment:
`_detect_source` should not be called again if there is already a translation for this message (we can just check for it, as it is stored inside it)

Reply:
I decided to slightly refactor the code to hide this inelegant logic away from the controller: 
https://github.com/odoo/odoo/pull/126555/files#diff-99e830e4c644c768bc059ee253cde05b242994b6f5266a0aae5ecf816045dffeR47

Note that I am not a huge fan of returning a tuple, maybe I could create two different methods instead.

Reply:
Indeed this API is a little confusing, we probably don't need a separate method for that, especially if it's only used once.

Reply:
Removed the dedicated method and added the logic in the _detect_source() method of the controller. Also improved the server side unit tests to ensure no extra API request is sent if the source exist in DB and ensure the translations are cleared correctly when the body of a message is edited.